src/azul/indexer/document.py

```diff
@@ -396,8 +400,18 @@ def translate_fields(cls,
                 return [cls.translate_fields(val, field_types, forward=forward, path=path) for val in doc]
             else:
                 assert False, (path, type(doc))
-        elif isinstance(field_types, FieldType):
-            field_type = field_types
+        else:
+            if isinstance(field_types, list):
+                # FIXME: Assert that a non-list field_type implies a non-list
+                #        doc (only possible for contributions).
+                #        https://github.com/DataBiosphere/azul/issues/2689
+                assert ((doc is None if forward else doc == NullableString.null_string)
```

Comment:
I find it odd that this mentions a specific null representation, those of nullable strings. I'm not even sure we support nullable lists. If you omit this term, does the assertion fire during unit or integration tests? If so where?

Reply:
You're right, that is odd. It should have raised a flag for me.

It causes some unit tests to fail.

One such failure is `test_hca_indexer.py`'s `test_cell_line_sample()`:
```
Traceback (most recent call last):
  File "/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/jesse/gi/azul/test/indexer/test_hca_indexer.py", line 1091, in test_cell_line_sample
    self._index_canned_bundle(bundle_fqid)
  File "/home/jesse/gi/azul/test/indexer/__init__.py", line 100, in _index_canned_bundle
    cls._index_bundle(bundle, delete=delete)
  File "/home/jesse/gi/azul/test/indexer/__init__.py", line 107, in _index_bundle
    cls.index_service.index(cls.catalog, bundle)
  File "/home/jesse/gi/azul/src/azul/indexer/index_service.py", line 147, in index
    self.aggregate(tallies)
  File "/home/jesse/gi/azul/src/azul/indexer/index_service.py", line 244, in aggregate
    contributions = self._read_contributions(total_tallies)
  File "/home/jesse/gi/azul/src/azul/indexer/index_service.py", line 362, in _read_contributions
    contributions = [
  File "/home/jesse/gi/azul/src/azul/indexer/index_service.py", line 363, in <listcomp>
    Contribution.from_index(self.catalogued_field_types(), hit)
  File "/home/jesse/gi/azul/src/azul/indexer/document.py", line 472, in from_index
    source = cls.translate_fields(hit['_source'], field_types[coordinates.entity.catalog], forward=False)
  File "/home/jesse/gi/azul/src/azul/indexer/document.py", line 393, in translate_fields
    new_doc[key] = cls.translate_fields(val, field_type, forward=forward, path=(*path, key))
  File "/home/jesse/gi/azul/src/azul/indexer/document.py", line 393, in translate_fields
    new_doc[key] = cls.translate_fields(val, field_type, forward=forward, path=(*path, key))
  File "/home/jesse/gi/azul/src/azul/indexer/document.py", line 400, in translate_fields
    return [cls.translate_fields(val, field_types, forward=forward, path=path) for val in doc]
  File "/home/jesse/gi/azul/src/azul/indexer/document.py", line 400, in <listcomp>
    return [cls.translate_fields(val, field_types, forward=forward, path=path) for val in doc]
  File "/home/jesse/gi/azul/src/azul/indexer/document.py", line 393, in translate_fields
    new_doc[key] = cls.translate_fields(val, field_type, forward=forward, path=(*path, key))
  File "/home/jesse/gi/azul/src/azul/indexer/document.py", line 408, in translate_fields
    assert doc is None or isinstance(doc, list)
AssertionError
```

Reply:
I will provide more info on what is doc and path. Possibly provide a solution.

Reply:
When this fails, doc is `~null`, field_types is `[NullableString]`, path is `contents.samples.disease`.

Reply:
The reason we get a None value even though, from field_types we expect at least a list of None is because we're looking at `samples`. Samples contain all of the fields for it's composite parts (so samples have a consistent document shape). If the type that samples represents doesn't contain a field (as is the case of disease) the field is stored as `None`.

In the particular case I was debugging, `Samples` was representing a `CellLine`, not a `Specimen` so the disease field was stored as `None`.

Reply:
One solution is to ignore the assertion of we're looking at a samples. I think the assert could then be changed to something like:
```python
assert is_samples() or isinstance(doc, list)
```
or even stronger:
```python
assert not is_samples() == isinstance(doc, list)
```
although I haven't tested that (and won't tonight :moon: )

Reply:
I confirmed this works, I'm pushing a version for review.